### PR TITLE
MacOS fixes

### PR DESCRIPTION
**Pull request by isuruf**
With these changes I could get 1.1 to compile, but fails at the link stage.
```
[ 97%] Linking CXX shared library libOpenCL.dylib
Undefined symbols for architecture x86_64:
  "_POclFinish", referenced from:
      _pocl_command_enqueue in pocl_util.c.o
ld: symbol(s) not found for architecture x86_64
```

There's a warning that it's undefined. Full log [here](https://travis-ci.org/isuruf/pocl-feedstock/jobs/352510601#L3024)
```
/Users/travis/miniconda3/conda-bld/pocl_1520884248289/work/pocl-1.1/lib/CL/pocl_util.c:533:5: warning: implicit declaration of function 'POclFinish' is invalid in C99 [-Wimplicit-function-declaration]
    POclFinish (command_queue);
    ^
```

Brief explanations are in commit messages.

cc @inducer

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -408,8 +408,10 @@ if(UNIX)
   include(CheckSymbolExists)
 
   # don't allow implicit function declarations
-  set(CMAKE_REQUIRED_FLAGS "-std=c99")
-  set(CMAKE_REQUIRED_LIBRARIES "rt")
+  set(CMAKE_REQUIRED_FLAGS "-std=c99 -Werror=implicit-function-declaration")
+  if (NOT APPLE)
```
Comment (franz) [inline CMakeLists.txt]: I'm thinking ``(NOT APPLE)`` can be changed to ``(CMAKE_SYSTEM_NAME MATCHES "Linux")`` since librt seems to be Linux-specific.

**Comment by franz** [comment]
> POclFinish (command_queue);

Yep that's a bug, it should be ``POname(clFinish) (command_queue);``

I have one small nitpick at the PR, otherwise LGTM.

**Comment by franz** [comment]
I see you fixed HAVE_CLOCK_GETTIME, thanks. FYI i've already rewritten the HAVE_FORK/VFORK tests to use CHECK_C_SOURCE_COMPILES so you don't need to fix those.

File: CMakeLists.txt [hunk 2]
```diff
@@ -408,18 +408,27 @@ if(UNIX)
...
-  CHECK_SYMBOL_EXISTS("clock_gettime"
-                      "time.h"
-                      HAVE_CLOCK_GETTIME)
+
+  CHECK_C_SOURCE_COMPILES("
+#include <time.h>
+
+int main(int argc, char** argv) {
+  struct timespec timeout = {0, 0};
+  clock_gettime(CLOCK_REALTIME, &timeout);
```
Comment (isuruf) [inline CMakeLists.txt]: Should this have `futimens(fd, NULL)`? There's code like
```
#ifdef HAVE_CLOCK_GETTIME
    futimens(fd, NULL);
#else
#endif
```
in LLVMFileUtils.cc

In the build that I said was successful I'm doing cmake `-DHAVE_CLOCK_GETTIME=0`

Comment (isuruf) [reply CMakeLists.txt]: Added a new check `HAVE_FUTIMENS`

**Comment by isuruf** [comment]
@franz, running the test suite revealed a few more bugs. Now all tests pass on macosx.

**Comment by franz** [comment]
```
/home/buildslave/basedir/quick_noLLVM_noDEBUG/source/lib/CL/devices/common.c:1031:1: warning: 
missing initializer for field ‘max_ever_allocated’ of ‘pocl_global_mem_t {aka struct pocl_global_mem_t}’ [-Wmissing-field-initializers]
 static pocl_global_mem_t system_memory = POCL_LOCK_INITIALIZER;
```

``pocl_global_mem_t`` is a struct, not a lock. Out of curiosity, is the pthread lock initializer on Mac OS X something else than zeroes ?

**Comment by isuruf** [comment]
How should I initialize `system_memory` variable? In OSX, `PTHREAD_MUTEX_INITIALIZER` is defined as below.
```
#define _PTHREAD_MUTEX_SIG_init          0x32AAABA7
#define PTHREAD_MUTEX_INITIALIZER {_PTHREAD_MUTEX_SIG_init, {0}}
```

**Comment by franz** [comment]
``system_memory =  { POCL_LOCK_INITIALIZER, 0, 0, 0 }`` should work, or you could initialize just the lock about 10 lines below, in the ``if (system_memory.total_alloc_limit == 0)`` block, since that block is only run once.

File: CMakeLists.txt [hunk 3]
```diff
@@ -408,18 +408,27 @@ if(UNIX)
   include(CheckSymbolExists)
 
   # don't allow implicit function declarations
-  set(CMAKE_REQUIRED_FLAGS "-std=c99")
-  set(CMAKE_REQUIRED_LIBRARIES "rt")
+  set(CMAKE_REQUIRED_FLAGS "-std=c99 -Werror=implicit-function-declaration")
```
Comment (pjaaskel) [inline CMakeLists.txt]: Is the -Werror switch something we can expect from ~all compilers and supported versions in UNIX?

Comment (franz) [reply CMakeLists.txt]: Good Q. At least clang & gcc should support it, dunno about others.

Comment (franz) [reply CMakeLists.txt]: @pjaaskel i'll take care of it in another commit (upcoming PR).

**Comment by oscarbg** [comment]
Hi @franz @isuruf,
just comment that seeing this pull I tested by myself also building it..
my findings are without reading the docs:
used precompiled llvm 6.0 from http://releases.llvm.org/6.0.0/clang+llvm-6.0.0-x86_64-apple-darwin.tar.xz and put in PATH although I remember now I also had Xcode 9.3 beta command line tools installed although I saw from CMake output used llvm-config and LLVM components from precompiled 6.0 folder .. 
1)needs -DENABLE_POCL_ICD=0 setting in cmake if not complains about no OSX support..
2)strangely had to include some #define HAVE_VFORK or similar as 
#define HAVE_VFORK 1
just before 
int pocl_run_command (char *const *args)
if not got #error Must have fork() or vfork() system calls for HSA
strange as I have ENABLE_HSA=0 ..
3)then applied patches as needed from here and compiled.. strangely didn't need to apply rt,-Werror=implicit-function-declaration and Improve HAVE_CLOCK_GETTIME test patches..
then builds..
( I also see WARNING: Linking two modules of different data layouts: '/Users/oscar/Downloads/pocl/build/lib/kernel/host/skylake/barrier.ll.bc' is '' whereas 'llvm-link' is 'e-m:o-i64:64-f80:128-n8:16:32:64-S128')
 but running example1:
I get:
Assertion failed: (r == 0), function void *pocl_memalign_alloc_global_mem(cl_device_id, size_t, size_t), file /Users/oscarbarenysgarcia/Downloads/pocl/lib/CL/devices/common.c, line 1170.
Abort trap: 6
as quick hack I comment  //POCL_LOCK (mem->pocl_lock); 
and // POCL_UNLOCK (mem->pocl_lock);
in pocl_memalign_alloc_global_mem and pocl_free_global_mem
but still get errors..
done late last night.. sorry in advance if results are my own errors as perhaps I did some step bad..

**Comment by franz** [comment]
@oscarbg 

1) i recommend to NOT use precompiled LLVM from llvm.org, unless you're **absolutely** sure it's was compiled on the same Mac OS X version as you're running. I have experience with random and mysterious linking errors when using precompiled LLVM from llvm.org. LLVM uses C++ heavily and compiled C++ is NOT portable across OS versions.

2) there's only ``-DENABLE_ICD=0`` not ``-DENABLE_POCL_ICD=0``

3) "strange as I have ENABLE_HSA=0" - i have no idea how you got that

4) you really want those patches, especially the Werror=implicit-function-declaration patch

5) "WARNING: Linking two modules of different data layouts" is harmless

6) "Assertion failed: (r == 0)" <<< this means one of the pthread functions returned non-zero result, which is  a serious problem. Commenting out POCL_LOCK/UNLOCK will really not fix anything, you'll have to find out what error it returns (print errno) and why.

In short: don't waste time with patching and dowloading LLVM. Clone this branch: https://github.com/franz/pocl/tree/travis-ci and take a look at last commit.

**Comment by isuruf** [comment]
@oscarbg, I fixed the assertion failed error in the last commit. Please use all the patches from here.

**Comment by oscarbg** [comment]
yes the assertion is now fixed @isuruf ,thanks!
example1 is now working correctly (prints some numbers &OK)

thanks @franz! ,
for all responses yeah was a -DENABLE_ICD=0 sorry I was speaking from memory..
seeing comments I will build LLVM by myself then and guess the only strange thing is the missing VFORK or FORK undefined but hope that will go away also..

just two questions:
1) this seems obvious: also now example1-spir is not working but it's because the LLVM used has to be built with support for it right?
as using the prebuilt one I see: 
./llvm-config --targets-built
AArch64 AMDGPU ARM BPF Hexagon Lanai Mips MSP430 NVPTX PowerPC Sparc SystemZ X86 XCore
missing SPIR..
2) all AVX512 support added recently is portable and should work out of the box on MacOS right?

will be nice of someone test vs Apple OpenCL driver now that they are shipping AVX512 enabled HW on Mac Pros to see if their OpenCL stack is making use of it..

**Comment by isuruf** [comment]
I also had the fork and vfork issue  and I fixed it, but I can't remember which patch fixed it. Did you use all the patches from this PR? Can you post the cmake output and CMakeFiles/*.log files somewhere and post  a link here?

**Comment by franz** [comment]
@oscarbg  AVX512 should be portable, but remember 1) use LLVM 6 because earlier have buggy AVX512, and 2) the vectorizer in LLVM is far from ideal, so you'll likely have to write hand-vectorized code using float16 or double8 vectors to see any AVX512.

**Comment by franz** [comment]
also, SPIR is not listed in the LLVM targets, because it's not a target. Pocl uses a test to find out SPIR support, you should see it somewhere in the CMake output.

**Comment by oscarbg** [comment]
thanks @franz..
@isuruf is the "Don't link to rt on macos" which solves the fork vfork issue.. thanks..
patches for me seem good enough to perhaps be merged after some cleanup..
